src/software/backend/simulation/physics_field.cpp

```diff
@@ -0,0 +1,110 @@
+#include "software/backend/simulation/physics_field.h"
+
+#include "software/backend/simulation/box2d_util.h"
+
+PhysicsField::PhysicsField(std::shared_ptr<b2World> world, const Field &field)
+    : field(field)
+{
+    setupFieldBoundary(world, field);
+    setupEnemyGoal(world, field);
+    setupFriendlyGoal(world, field);
+}
+
+void PhysicsField::setupFieldBoundary(std::shared_ptr<b2World> world, const Field &field)
+{
+    field_boundary_body_def.type = b2_staticBody;
+    // Note that the body shape is defined relative to the body position. Setting the
+    // body position to (0, 0) makes it easy to add the shape using the standard field
+    // interface, since field coordinates are also relative to (0, 0)
+    // https://www.iforce2d.net/b2dtut/fixtures
+    field_boundary_body_def.position.Set(0, 0);
+    field_boundary_body = world->CreateBody(&field_boundary_body_def);
+
+    const unsigned int shape_loop_size = 4;
+    b2Vec2 vs[shape_loop_size]         = {
+        createVec2(field.fieldBoundary().posXPosYCorner()),
+        createVec2(field.fieldBoundary().posXNegYCorner()),
+        createVec2(field.fieldBoundary().negXNegYCorner()),
+        createVec2(field.fieldBoundary().negXPosYCorner()),
+    };
+    field_boundary_shape.CreateLoop(vs, shape_loop_size);
+    field_boundary_fixture_def.shape = &field_boundary_shape;
+    // Collisions with the field boundary are perfectly elastic and have
+    // no friction
+    field_boundary_fixture_def.restitution = 1.0;
+    field_boundary_fixture_def.friction    = 0.0;
+    field_boundary_body->CreateFixture(&field_boundary_fixture_def);
+}
+
+void PhysicsField::setupEnemyGoal(std::shared_ptr<b2World> world, const Field &field)
+{
+    enemy_goal_body_def.type = b2_staticBody;
+    // Note that the body shape is defined relative to the body position. Setting the
+    // body position to (0, 0) makes it easy to add the shape using the standard field
+    // interface, since field coordinates are also relative to (0, 0)
+    // https://www.iforce2d.net/b2dtut/fixtures
+    enemy_goal_body_def.position.Set(0, 0);
+    enemy_goal_body = world->CreateBody(&enemy_goal_body_def);
+
+    const unsigned int shape_chain_size          = 4;
```

Comment:
suggest changing to `num_enemy_goal_vertices`?

Reply:
fixed